File: hw-6B/Python/C-cubic-equation.py
def equation(params, x):
    a, b, c, d = params
    return a * x**3 + b * x**2 + c * x + d


def check_pos(m, params):
    return equation(params, m) > 0


def check_neg(m, params):
    return equation(params, m) < 0


def left_binsearch(check, l, r, params, eps=10e-6):
    while l + eps < r:
        m = (r + l) / 2.0
        if check(m, params):
            r = m
        else:
            l = m
    return l


if __name__ == '__main__':

    """Solve cubic equation"""

    eps = 2e-7
    coef = [int(x) for x in input().split()]
    if coef[0] > 0:
        ans = left_binsearch(check_pos, -1001, 1001, coef, eps)
    else:
        ans = left_binsearch(check_neg, -1001, 1001, coef, eps)
    print(ans)

File: hw-6B/C++/C-cubic-equation.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <iomanip>

using namespace std;

double equation(int coef[], double x){
    return coef[0] * pow(x, 3) + coef[1] * pow(x, 2) + coef[2] * x + coef[3];
}

bool checkPos(int coef[], double m){
    return equation(coef, m) > 0;
}

bool checkNeg(int coef[], double m){
    return equation(coef, m) < 0;
}

double leftBinSearch(bool (*check)(int[], double), int coef[], double l, double r, double eps){
    
    double m;
    while (l + eps < r){
        m = (l + r) / 2.0;
        if (check(coef, m))
            r = m;
        else
            l = m;
    }
    return l;

}

int main(){

    // Solve cubic equation
 
    int coef[4];
    double ans;
    for (int i=0; i<4; i++){
        cin >> coef[i];
    }
    double eps = 1e-6;
    if (coef[0] > 0)
        ans = leftBinSearch(&checkPos, coef, -1001., 1001., eps);
    else
        ans = leftBinSearch(&checkNeg, coef, -1001., 1001., eps);
    
    cout << setprecision(15) << ans << "\n";
   
    return 0;
}